### DISCUSSION TITLE
What's the logic for permitting loading an ESM module but not a static file in middleware?

### DISCUSSION THREAD

**Discussion by motdotla** [vercel/next.js#38088] [Help]
What's the logic that does not permit loading a static file in vercel middleware but does permit loading an ESM module? 

Wouldn't an ESM module potentially have more risks than loading a static file that is part of committed source code?

**Comment by GabenGar** [comment]
It's not about security, reading files requires a nodejs module to which middlewares have no access.

**Reply by motdotla** [reply]
I understand that, but my question is why -  what's the logic for that forced decision?

It certainly has a file system on it. I assume this is deployed on AWS Lambda - to which the best of my knowledge I am still able to use the fs library on - reading any local filed bundled as part of my lambda deploy.

**Reply by GabenGar** [reply]
The logic is not to have 200mb+ lambdas.

**Reply by motdotla** [reply]
> The logic is not to have 200mb+ lambdas.

That doesn't make sense. By that same logic I could just as easily embed a 200mb json file inside an ES module.

**Reply by GabenGar** [reply]
What doesn't make sense? You can't read files without access to nodejs API, simple as.

**Reply by motdotla** [reply]
Ok, so maybe I'm confused on what this middleware is. My understanding is they are deployed on node infrastructure - I think on AWS Lambda. Are they just front-end code deployed to the browser? Or front-end code deployed to headless browsers on the 'edge' networks?

**Reply by GabenGar** [reply]
No, they are the server-only code deployed on edge infra and have access to [select few web interfaces](https://nextjs.org/docs/api-reference/edge-runtime#unsupported-apis). That in vercel case the infra is ran atop of aws is just an implementation detail.
